File: app/src/main/java/com/example/android/weeklytips/dialogs/EditTextDialog.java
package com.example.android.weeklytips.dialogs;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.example.android.weeklytips.R;

public class EditTextDialog extends DialogFragment {

    private static final String PROMPT_KEY = "prompt_key";
    private static final String VALUE_KEY = "value_key";

    private EditTextDialogListener mListener;
    private EditText etValue;

    public static EditTextDialog newInstance(String prompt, String currentValue) {

        Bundle args = new Bundle();
        args.putString(PROMPT_KEY, prompt);
        args.putString(VALUE_KEY, currentValue);

        EditTextDialog fragment = new EditTextDialog();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof EditTextDialogListener) {
            mListener = (EditTextDialogListener) context;
        } else {
            throw new ClassCastException("Caller must implement Listener");
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.dialog_edit_text, container, false);

        String prompt = getArguments().getString(PROMPT_KEY);
        String initialValue = getArguments().getString(VALUE_KEY);
        if (initialValue == null) {
            initialValue = "";
        }

        TextView tvPrompt = (TextView) view.findViewById(R.id.tvPrompt);
        tvPrompt.setText(prompt);

        etValue = (EditText) view.findViewById(R.id.etValue);
        etValue.setText(initialValue);
        etValue.setSelection(etValue.length());

        Button btnOk = (Button) view.findViewById(R.id.btnOk);
        final String finalInitialValue = initialValue;
        btnOk.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String newValue = etValue.getText().toString();
                if (!newValue.equals(finalInitialValue)) {
                    mListener.onEditTextDialogOK(newValue, getTag());
                }
                dismiss();
            }
        });

        Button btnCancel = (Button) view.findViewById(R.id.btnCancel);
        btnCancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });

        return view;
    }

    public interface EditTextDialogListener {
        void onEditTextDialogOK(String newValue, String tag);
    }

}

File: app/src/main/java/com/example/android/weeklytips/dialogs/EditTextDialog.kt
package com.example.android.weeklytips.dialogs

import android.content.Context
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView

import com.example.android.weeklytips.R

class EditTextDialog : DialogFragment() {

    private var mListener: EditTextDialogListener? = null
    private var etValue: EditText? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is EditTextDialogListener) {
            mListener = context as EditTextDialogListener?
        } else {
            throw ClassCastException("Caller must implement Listener")
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val view = inflater!!.inflate(R.layout.dialog_edit_text, container, false)

        val prompt = arguments.getString(PROMPT_KEY)
        var initialValue = arguments.getString(VALUE_KEY)
        if (initialValue == null) {
            initialValue = ""
        }

        val tvPrompt = view.findViewById(R.id.tvPrompt) as TextView
        tvPrompt.text = prompt

        etValue = view.findViewById(R.id.etValue) as EditText
        etValue!!.setText(initialValue)
        etValue!!.setSelection(etValue!!.length())

        val btnOk = view.findViewById(R.id.btnOk) as Button
        val finalInitialValue = initialValue
        btnOk.setOnClickListener {
            val newValue = etValue!!.text.toString()
            if (newValue != finalInitialValue) {
                mListener!!.onEditTextDialogOK(newValue, tag)
            }
            dismiss()
        }

        val btnCancel = view.findViewById(R.id.btnCancel) as Button
        btnCancel.setOnClickListener { dismiss() }

        return view
    }

    interface EditTextDialogListener {
        fun onEditTextDialogOK(newValue: String, tag: String)
    }

    companion object {

        private val PROMPT_KEY = "prompt_key"
        private val VALUE_KEY = "value_key"

        fun newInstance(prompt: String, currentValue: String): EditTextDialog {

            val args = Bundle()
            args.putString(PROMPT_KEY, prompt)
            args.putString(VALUE_KEY, currentValue)

            val fragment = EditTextDialog()
            fragment.arguments = args
            return fragment
        }
    }

}